dvc/fs/repo.py

```diff
@@ -340,18 +329,34 @@ def walk(
                 onerror(NotADirectoryError(top))
             return
 
-        ignore_subrepos = not self._traverse_subrepos
-        if follow_subrepos is not None:
-            ignore_subrepos = not follow_subrepos
+        repo = self._get_repo(os.path.abspath(top))
+        ignore_subrepos = kwargs.pop("ignore_subrepos", True)
+        if repo:
+            yield from repo.dvcignore(
+                self.walk_fs(
+                    top,
+                    topdown=topdown,
+                    onerror=onerror,
+                    dvcfiles=dvcfiles,
+                    **kwargs
+                ),
+                ignore_subrepos=ignore_subrepos,
+            )
+        else:
+            yield from self.walk_fs(
+                top,
+                topdown=topdown,
+                onerror=onerror,
+                dvcfiles=dvcfiles,
+                **kwargs
+            )
 
+    def walk_fs(
```

Comment:
No need to have it as a public method.

```suggestion
    def _walk_fs(
```

Or there might be a better name too, not sure.